### PR TITLE
Issue #2504031: Fix __construct of plugin to be compatible with D8 Plugins

### PR DESCRIPTION
**Pull request by drupol**
Update code to be compatible with D8.

### CODE REVIEW

File: src/DependencyInjection/Container.php [hunk 1]
```diff
@@ -228,6 +230,27 @@ public function has($id) {
   /**
    * {@inheritdoc}
    */
+  public function createInstance($plugin_id, $configuration) {
+    $service_definition = $this->getDefinition($plugin_id);
+    $service_definition += array(
+      'arguments' => array()
+    );
+
+    array_unshift($service_definition['arguments'], $configuration);
```
Comment (fabianx-ai) [inline src/DependencyInjection/Container.php]: Lets move all of this into the caller (ContainerAwarePluginManager::createInstance)

File: src/DependencyInjection/Container.php [hunk 2]
```diff
@@ -228,6 +230,27 @@ public function has($id) {
   /**
    * {@inheritdoc}
    */
+  public function createInstance($plugin_id, $configuration) {
```
Comment (fabianx-ai) [inline src/DependencyInjection/Container.php]: Let this take a $service_definition argument, please.

Comment (fabianx-ai) [inline src/DependencyInjection/Container.php]: so:

($plugin_id, $service_definition)

**Comment by dawehner** [comment]
Mh I think its odd to expose the plugin information on a container

File: src/Plugin/ContainerAwarePluginManager.php [hunk 3]
```diff
@@ -61,8 +61,7 @@ public function hasDefinition($plugin_id) {
   public function createInstance($plugin_id, array $configuration = array()) {
     // @todo: Use ->expandArguments() when get() disallows getting private
     //        services.
-    $plugin = clone $this->container->get($this->servicePrefix . $plugin_id);
-    return $plugin;
+    return $this->container->createInstance($this->servicePrefix . $plugin_id, $configuration);
```
Comment (fabianx-ai) [inline src/Plugin/ContainerAwarePluginManager.php]: Call this with the $definition instead, get the definition via: $this->getDefinition($plugin_id);

**Comment by drupol** [comment]
Can't get this working.

File: src/DependencyInjection/Container.php [hunk 4]
```diff
@@ -218,6 +218,8 @@ public function set($id, $service, $scope = self::SCOPE_CONTAINER) {
     $this->services[$id] = $service;
   }
 
+
+
```
Comment (fabianx-ai) [inline src/DependencyInjection/Container.php]: white space change

File: src/Plugin/ContainerAwarePluginManager.php [hunk 5]
```diff
@@ -61,8 +61,14 @@ public function hasDefinition($plugin_id) {
   public function createInstance($plugin_id, array $configuration = array()) {
     // @todo: Use ->expandArguments() when get() disallows getting private
     //        services.
-    $plugin = clone $this->container->get($this->servicePrefix . $plugin_id);
-    return $plugin;
+    $service_definition = $this->getDefinition($plugin_id);
+    $service_definition += array(
+      'arguments' => array()
```
Comment (fabianx-ai) [inline src/Plugin/ContainerAwarePluginManager.php]: nit - Missing ',' at end of line

File: src/Plugin/ContainerAwarePluginManager.php [hunk 6]
```diff
@@ -61,8 +61,14 @@ public function hasDefinition($plugin_id) {
   public function createInstance($plugin_id, array $configuration = array()) {
     // @todo: Use ->expandArguments() when get() disallows getting private
     //        services.
-    $plugin = clone $this->container->get($this->servicePrefix . $plugin_id);
-    return $plugin;
+    $service_definition = $this->getDefinition($plugin_id);
```
Comment (fabianx-ai) [inline src/Plugin/ContainerAwarePluginManager.php]: Lets call this plugin_definition.

File: src/Plugin/ContainerAwarePluginManager.php [hunk 7]
```diff
@@ -61,8 +61,14 @@ public function hasDefinition($plugin_id) {
   public function createInstance($plugin_id, array $configuration = array()) {
     // @todo: Use ->expandArguments() when get() disallows getting private
     //        services.
```
Comment (fabianx-ai) [inline src/Plugin/ContainerAwarePluginManager.php]: nit - Lets remove those two lines.

**Comment by fabianx-ai** [comment]
@dawehner: It is, the definitions should be parameters per plugin manager. (and yes I want that info in the container for now for those plugin managers).

This is a first step into that direction, createInstance is only temporary. I think the real fix is to have get() also take a definition array.

**Comment by fabianx-ai** [comment]
@drupol The code is great. The unit tests fail, because of changed expectations. I will fix them - likely tomorrow.

**Comment by drupol** [comment]
Ok, I will look for it in the meantime ! Thanks for your patience :)

**Comment by drupol** [comment]
If it helps, when I try to run Openlayers with these changes, I get the error:

```
RuntimeException: The "plugin_openlayers.Map.internal.OLMap" service definition does not exist. in Drupal\service_container\DependencyInjection\Container->get() (line 88 of /home/pol/git/service_container_fork/src/DependencyInjection/Container.php).
```

File: src/Plugin/ContainerAwarePluginManager.php [hunk 8]
```diff
@@ -59,10 +59,14 @@ public function hasDefinition($plugin_id) {
...
-    $plugin = clone $this->container->get($this->servicePrefix . $plugin_id);
-    return $plugin;
+    $plugin_definition = $this->getDefinition($plugin_id);
+    $plugin_definition += array(
+      'arguments' => array(),
+    );
+
+    array_unshift($plugin_definition['arguments'], $configuration);
+
+    return $this->container->createInstance($this->servicePrefix . $plugin_id, $configuration);
```
Comment (fabianx-ai) [inline src/Plugin/ContainerAwarePluginManager.php]: DOH!

Comment (fabianx-ai) [inline src/Plugin/ContainerAwarePluginManager.php]: I see the bug now, need to pass $plugin_definition and not $configuration here ...

Comment (fabianx-ai) [inline src/Plugin/ContainerAwarePluginManager.php]: @drupol ^^

**Comment by drupol** [comment]
Now I don't have any errors and it seems that it has fixed a couple of things. BUT... I have now plenty of:

```
Warning: Missing argument 3 for Drupal\openlayers\Types\Object::__construct(), called in /home/pol/git/service_container_fork/src/DependencyInjection/Container.php on line 154 and defined in Drupal\openlayers\Types\Object->__construct() (line 88 of /home/pol/git/openlayers/src/Types/Object.php).
Notice: Undefined index: plugin module in Drupal\openlayers\Types\Object->__construct() (line 90 of /home/pol/git/openlayers/src/Types/Object.php).
Notice: Undefined index: plugin type in Drupal\openlayers\Types\Object->__construct() (line 90 of /home/pol/git/openlayers/src/Types/Object.php).
Notice: Undefined index: name in Drupal\openlayers\Types\Object->__construct() (line 90 of /home/pol/git/openlayers/src/Types/Object.php).
Warning: Missing argument 3 for Drupal\openlayers\Types\Object::__construct(), called in /home/pol/git/service_container_fork/src/DependencyInjection/Container.php on line 154 and defined in Drupal\openlayers\Types\Object->__construct() (line 88 of /home/pol/git/openlayers/src/Types/Object.php).
```

I will look for them at home.

**Comment by fabianx-ai** [comment]
@drupol Yes, that is the BC break I talked about.

We should break BC some more before you fix it ;).

**Comment by drupol** [comment]
No problem :)

**Comment by fabianx-ai** [comment]
To break BC some more we should use:
- $configuration
- $plugin_id
- $plugin_definition (already in arguments as first part)
- any further parameters

As the first parameters.

That means array_unshift the plugin_id first and then the configuration.

Great work!

**Comment by drupol** [comment]
It works if I do some change in the constructor of my OL objects:

```
  /**
   * Constructs a Drupal\Component\Plugin\PluginBase object.
   *
   * @param array $configuration
   *   A configuration array containing information about the plugin instance.
   */
  public function __construct(array $configuration, $plugin_id, array $plugin_definition) {
    $this->pluginDefinition = $plugin_definition;

    $this->pluginId = $plugin_definition['plugin module'] . '.' . $plugin_definition['plugin type'] . ':' . $plugin_definition['name'];
    $this->configuration = $configuration;
  }
```

But $configuration is always empty.

**Comment by drupol** [comment]
Actually, I can now remove the constructor method from Objects.php ! :+1: 
I don't need the pluginId property.

**Comment by fabianx-ai** [comment]
@drupol We will need a huge change notice both for service_container and openlayers.

I will also have to fix render_cache first.

**Comment by drupol** [comment]
Huge ?
I can take care of it if you give me more informations. Right now, things hasn't changed much in Openlayers, why would it be so huge ? I don't see what you mean.

**Comment by fabianx-ai** [comment]
Triggering a rebuild.

**Pull request merged**